types/time.go

```diff
@@ -207,159 +184,191 @@ func FromGoTime(t gotime.Time) MysqlTime {
+		return ZeroCoreTime, false
+	}
+	if uint64(day) >= (1 << (dayBitFieldStart - dayBitFieldEnd + 1)) {
+		return ZeroCoreTime, false
+	}
+	if uint64(hour) >= (1 << (hourBitFieldStart - hourBitFieldEnd + 1)) {
+		return ZeroCoreTime, false
+	}
+	if uint64(minute) >= (1 << (minuteBitFieldStart - minuteBitFieldEnd + 1)) {
+		return ZeroCoreTime, false
+	}
+	if uint64(second) >= (1 << (secondBitFieldStart - secondBitFieldEnd + 1)) {
+		return ZeroCoreTime, false
+	}
+	if uint64(microsecond) >= (1 << (microsecondBitFieldStart - microsecondBitFieldEnd + 1)) {
+		return ZeroCoreTime, false
+	}
+	return FromDate(year, month, day, hour, minute, second, microsecond), true
 }
 
-// Clock returns the hour, minute, and second within the day specified by t.
-func (t Time) Clock() (hour int, minute int, second int) {
-	return t.time.Hour(), t.time.Minute(), t.time.Second()
-}
+// coreTime is an alias to CoreTime, embedd in Time.
+type coreTime = CoreTime
 
 // Time is the struct for handling datetime, timestamp and date.
-// TODO: check if need a NewTime function to set Fsp default value?
 type Time struct {
-	time MysqlTime
-	tp   uint8
-	// Fsp is short for Fractional Seconds Precision.
-	// See http://dev.mysql.com/doc/refman/5.7/en/fractional-seconds.html
-	fsp int8
+	coreTime
 }
 
-// NewTime constructs time from core time, type and fsp.
-func NewTime(mt MysqlTime, tp uint8, fsp int8) Time {
-	return Time{
-		time: mt,
-		tp:   tp,
-		fsp:  fsp,
-	}
-}
-
-// Year returns year value.
-func (t Time) Year() int {
-	return t.time.Year()
-}
-
-// Month returns month value.
-func (t Time) Month() int {
-	return t.time.Month()
-}
-
-// Day returns day value.
-func (t Time) Day() int {
-	return t.time.Day()
+// Clock returns the hour, minute, and second within the day specified by t.
+func (t Time) Clock() (hour int, minute int, second int) {
+	return t.Hour(), t.Minute(), t.Second()
 }
 
-// Hour returns hour value.
-func (t Time) Hour() int {
-	return t.time.Hour()
-}
+const (
+	// Core time bit fields.
+	yearBitFieldStart, yearBitFieldEnd               uint64 = 63, 50
+	monthBitFieldStart, monthBitFieldEnd             uint64 = 49, 46
+	dayBitFieldStart, dayBitFieldEnd                 uint64 = 45, 41
+	hourBitFieldStart, hourBitFieldEnd               uint64 = 40, 36
+	minuteBitFieldStart, minuteBitFieldEnd           uint64 = 35, 30
+	secondBitFieldStart, secondBitFieldEnd           uint64 = 29, 24
+	microsecondBitFieldStart, microsecondBitFieldEnd uint64 = 23, 4
+
+	// fspTt bit field.
+	// `fspTt` format:
+	// | fsp: 3 bits | type: 1 bit |
+	// When `fsp` is valid (in range [0, 6]):
+	// 1. `type` bit 0 represent `DateTime`
+	// 2. `type` bit 1 represent `Timestamp`
+	//
+	// Since s`Date` does not require `fsp`, we could use `fspTt` == 0b1110 to represent it.
+	fspTtBitFieldStart, fspTtBitFieldEnd uint64 = 3, 0
+
+	yearBitFieldMask        uint64 = 0b1111111111111100000000000000000000000000000000000000000000000000
```

Comment:
I think it's not a good idea to use binary literal, how about `(1 << yearBitWidth - 1) << yearBitFieldEnd`